### PR TITLE
fix(hybrid-cloud): Adds an optional DB param to relay project cache on-commit logic

### PR DESCRIPTION
**Pull request by GabeVillalobos**
Addresses an issue with the `schedule_invalidate_project_config` method can pass an incorrect `using` parameter when creating a `transaction.on_commit` hook. I've added this as an optional parameter to allow the caller to determine the correct DB transaction to associate the request with, as it can vary by callsite. Not doing so will result in the `on_commit` handler executing _immediately_, which can cause cross database accesses within a single transaction.

### CODE REVIEW

**Review by iker-barriocanal — COMMENTED** [review]
Hi Gabe! I'm not sure if I fully understand the usage of this new parameter:

1. As a developer using this function, when/how do I know whether I should use the default database? How do I know which databases should I pick?
1. The code assumes an entire organization is stored in a single database, is this assumption correct? Will it hold in the short to medium future?

Personally, I'd benefit from finding the answers to these in the documentation of the `transaction_db` param.

**Review by jjbayer — CHANGES_REQUESTED** [review]
@GabeVillalobos are there any specific calls that should set this new parameter? Should they also be included in this PR? Any chance you could reproduce the failing behavior in a test case?

**Comment by GabeVillalobos** [comment]
> Hi Gabe! I'm not sure if I fully understand the usage of this new parameter:
> 
> 1. As a developer using this function, when/how do I know whether I should use the default database? How do I know which databases should I pick?
> 2. The code assumes an entire organization is stored in a single database, is this assumption correct? Will it hold in the short to medium future?
> 
> Personally, I'd benefit from finding the answers to these in the documentation of the `transaction_db` param.

Thanks for the feedback, I'll update the documentation around this parameter to be more clear about its usage.

I think the source of confusion here is that this parameter does not decide where the data is going to be committed. Its purpose is to associate the `on_commit` handler with the correct surrounding transaction which only the caller will know.

The reason I'm adding this parameter is due to a usage of this function that I found [in getsentry](https://github.com/getsentry/getsentry/blob/25583be857fac4db761674282ec6a07291ca939b/getsentry/receivers/usage.py#L284) which invokes it in a post-save handler for an entity in a different DB. Because some of these calls are in transactions to a different database, the `on_commit` handler immediately fires and commits the work mid-transaction which does not _seem_ to be the correct intended behavior.

The current behavior will always provide `default` as the on_commit transaction DB which should be consistent with the default I've selected here.

**TLDR: The `transaction_db` parameter should be identical to the `using=` parameter of the surrounding transaction if one exists.**

A correct usage of this would be:
```python
transaction.atomic(using=router.db_for_write(Subscription)):
    # <db work here>
    schedule_invalidate_project_config(
        trigger="trigger_str",
        transaction_db=router.db_for_write(Subscription)
    ) 
```

@jjbayer this really should only affect getsentry code for now, but _may_ have some effect in our upcoming database split work so I'll audit any sentry repo usages of this and will post a follow-up PR with the correct `transaction_db` params if they are needed. As for testing, the existing tests should be sufficient, and I'll be adding tests that exercise this new functionality in getsentry/getsentry#11466.

File: src/sentry/tasks/relay.py [hunk 1]
```diff
@@ -251,6 +251,7 @@ def schedule_invalidate_project_config(
     project_id=None,
     public_key=None,
     countdown=5,
+    transaction_db="",
```
[the remark this answers is not in the document]
Comment (jjbayer) [reply src/sentry/tasks/relay.py:254]: I think it has to be `None` when we pass it into `on_commit`, because that function in turn calls `get_connection`, which expects `None` if I'm not mistaken:

https://github.com/django/django/blob/9cc0d7f7f85cecc3ad15bbc471fe6a08e4f515b6/django/db/transaction.py#L23-L24

**Review by jjbayer — COMMENTED** [review]
> A correct usage of this would be:
> 
> ```python
> transaction.atomic(using=router.db_for_write(Subscription)):
>     # <db work here>
>     schedule_invalidate_project_config(
>         trigger="trigger_str",
>         transaction_db=router.db_for_write(Subscription)
>     ) 
> ```

@GabeVillalobos makes sense to me, though I'm suprised that Django does not provide a `transaction.current` helper. That would prevent having to pass the `using` manually.

File: src/sentry/tasks/relay.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (GabeVillalobos) [reply src/sentry/tasks/relay.py:254]: Passing `None` to `on_commit` will result in an exception being thrown since [we've patched](https://github.com/getsentry/sentry/blob/8deb98a50929293c9973e6da39b31164d851e67a/src/sentry/silo/patches/silo_aware_transaction_patch.py#L54) `transaction.get_connection` to expect an explicit `using` parameter now. This is why we fallback to using the DB backing the `Project` model instead as a safe default.

**Review by dashed — APPROVED** [review]

**Comment by GabeVillalobos** [comment]
>@GabeVillalobos makes sense to me, though I'm suprised that Django does not provide a transaction.current helper. That would prevent having to pass the using manually.

@jjbayer we can have multiple active transactions to different databases open simultaneously in the same code path (and in some cases already do in getsentry), which is why this isn't easy to automate. We've begun introducing [testing utilities](https://github.com/getsentry/sentry/blob/8deb98a50929293c9973e6da39b31164d851e67a/src/sentry/testutils/hybrid_cloud.py#L125) to prevent this from happening, but likely can't assume which transaction to use until this is enforced globally.

**Review by jjbayer — APPROVED** [review]
